add graphql-authz blog post (#525)
the-guild-org/the-guild-website: pages/blog/graphql-authz.mdx

## Reducing database calls with post-execution rules

Querying data sources from authorization rules is powerful but it adds workload to data sources and, in most cases, the data needed to perform authorization is closely related to requested entities. There is a possibility to reduce data sources workload by performing data-dependent authorization logic in the post-execution phase.

With GraphQL AuthZ it’s possible for post-execution rules to require the data, related to the rule target, that is necessary for the rule to work. For example, if the rule is attached to some Object Field it could require any information about sibling fields with their relations. Another example is if the rule is attached to Object Type it could require any fields of the object with their relations.

```typescript
// creating post-execution rule

const CanReadPost = postExecRule({
  selectionSet: '{ status author { id } }',
})(
  (context, fieldArgs, post, parent) =>
    post.status === 'public' || post.author.id === context.user?.id
);
```

> This technique helps to avoid additional calls to data sources by executing authorization logic on top of the entire operation result with the related extra data being populated. Since related data is often stored in the same place it can be fetched from a data source in one call instead of making one call for authorization and one call for actual data populating.

Reviewer: ```suggestion
By using this technique we reduce remote procedure calls to individual data sources by executing authorization logic on top of GraphQL execution result that is enriched with the additional data specified by the authorization rules. Since related data is often stored in the same place it can be fetched from a data source in one roundtrip (via the GraphQL schema) instead of performing one remote procedure call for the authorization and one call for the actual data populating.
```

## Microservices

> With GraphQL AuthZ it is possible to implement a centralized gateway authorization layer as well as microservice level authorization. You can choose between storing the whole authorization schema in a holistic way in a gateway independent of GraphQL schema or let GraphQL schema manage authorization as part of itself.

Reviewer: ```suggestion
With GraphQL AuthZ it is possible to implement a centralized gateway authorization layer as well as microservice level authorization. You can choose between storing the whole authorization schema in a holistic way on a stitched or federated gateway or having dedicated authorization schemas for each subgraph/schema specified by your services. It is even possible to mix both approaches!
```

> “Shifting this configuration out of the gateway makes subschemas autonomous, and allows them to push their own configuration up to the gateway—enabling more sophisticated schema releases.” — [schema stitching handbook](https://github.com/gmac/schema-stitching-handbook/tree/master/stitching-directives-sdl)

> The `@graphql-authz/directive` package provides an importable directive that can be used to annotate types and fields within subschemas and a configuration transformer that can be used on the gateway to convert the subschema directives into explicit authorization settings.

Reviewer: ```suggestion
The `@graphql-authz/directive` package provides a GraphQL directive that can be used to annotate types and fields within your subschemas SDL and a configuration transformer that can be used on the gateway to convert the subschema directives into explicit authorization settings.
```

```graphql
# using @authz directive

type User {
  id: ID!
  email: String! @authz(rules: [IsAdmin])
  posts: [Post!]!
}

type Post @authz(rules: [CanReadPost]) {
  id: ID!
  title: String!
  body: String!
  status: Status!
  author: User!
}

type Query {
  users: [User!]! @authz(rules: [IsAuthenticated])
  post(id: ID!): Post
}

type Mutation {
  publishPost(postId: ID!): Post! @authz(rules: [CanPublishPost])
}
```

> Directives could be used in monolith architecture as well.

Reviewer: ```suggestion
On top of that directives can as well be used in monolith architecture.
```

> If directives are not applicable for some reason, plain JSON auth schema could be used to describe which rules should be executed for certain Object, Interface, Field, Query, or Mutation

Reviewer: ```suggestion
If you are not pursuing a schema-first (SDL) development flow and are more a fan of the code-first approach (which does not let you specify directives), the authorization schema can also be described as a plain JSON object. Allowing you to specify the rules that should be executed on your object types, interfaces, and fields.
```

Author: disagree here about the 
> code-first approach (which does not let you specify directives)

it's possible to specify directives with a code-first approach, there is an example with NestJS: https://github.com/AstrumU/graphql-authz/blob/main/examples/packages/nestjs/src/index.ts#L111

Reviewer: Interesting, I haven't seen this before. When using graphql-js API for building an executable schema using code this is not possible. So nestjs is definitely doing some framework magic on top of that. Maybe we should change `(which does not let you specify directives)` to `(which does not let you specify directives without framework-specific voodoo magic)`? :D

In another library I have built, I allowed mirroring directive functionality through extension fields see https://github.com/n1ru4l/graphql-public-schema-filter#usage-instructions

Author: Yeah, agree, I added this :) 

https://github.com/n1ru4l/graphql-public-schema-filter is a nice tool :) yeah, in graphql-authz I also allowed putting auth configuration right to extensions

```typescript
// defining auth schema

const authSchema = {
  Post: { __authz: { rules: ['CanReadPost'] } },
  User: {
    email: { __authz: { rules: ['IsAdmin'] } },
  },
  Mutation: {
    publishPost: { __authz: { rules: ['CanPublishPost'] } },
  },
  Query: {
    users: { __authz: { rules: ['IsAuthenticated'] } },
  },
  // wildcards are supported. read more: https://github.com/AstrumU/graphql-authz#wildcard-rules
  '*': {
    __authz: { rules: ['Reject'] },
    '*': {
      __authz: { rules: ['IsAuthenticated'] },
    },
  },
};
```

> ## Comparison with GraphQL Shield

Reviewer: ```suggestion
## Comparing GraphQL AuthZ with GraphQL Shield
```

> [GraphQL Shield](https://github.com/maticzav/graphql-shield) is a great tool for creating authorization layers. Actually, GraphQL AuthZ was inspired by GraphQL Shield! However, it incorporates a different approach compared to GraphQL AuthZ in terms of hooking up authorization logic to the execution process.

Reviewer: ```suggestion
[GraphQL Shield](https://github.com/maticzav/graphql-shield) is a great tool for creating authorization layers that has vast adoption from the community. In fact, GraphQL AuthZ is highly inspired by GraphQL Shield! However, GraphQL Shield uses a different approach compared to GraphQL AuthZ for applying authorization rules.
```

> The main difference is GraphQL Shield uses field middlewares to execute authorization logic while GraphQL AuthZ wraps the entire execution process. The benefits of the wrapping approach are described in previous paragraphs however there are some drawbacks. For example, with GraphQL AuthZ post-execution rules there is no ability to fail the request earlier because post-execution rules are executed after all resolvers are executed. GraphQL Shield, on the other hand, can fail the request during the execution phase which is earlier than the post-execution phase but later than the pre-execution phase. However post-execution rules have the benefit of accessing results of all field resolvers by design. It means that the post-execution rule attached to a certain field has an access to sibling fields even if these fields are resolved by field resolvers. Also, the post-execution rule attached to a certain field which is resolved by the field resolver has an access to the field value. The middleware approach is missing these possibilities because authorization logic is executed in the context of field resolvers and has access only to the parent object which is the result of the parent field resolver or query. Results of field resolvers of the current and sibling fields are not available to authorization logic at this moment.

Reviewer: ```suggestion
The main difference is GraphQL Shield uses field middleware by wrapping all the `resolver` functions within your GraphQL schema for executing authorization logic during the data-resolving phase, while GraphQL AuthZ wraps the entire `execute` logic.

The benefits of the wrapping approach are described in previous paragraphs, however, there are also some drawbacks. For example, with GraphQL AuthZ post-execution rules there is no ability to fail the request early because post-execution rules are executed after all resolvers are executed. GraphQL Shield, on the other hand, can fail the request during the execution phase which happens before the post-execution phase but later than the pre-execution phase. On the other hand, post-execution rules have the benefit of accessing the resolved value of the field they are specified on and, furthermore, even the sibling fields that are specified via the rules selection set.

The middleware approach of GraphQL shield is missing these possibilities because authorization logic is executed in the context of field resolvers and has access only to the parent object which is the value returned from the parent field resolver. At that stage, the wrapped field resolvers have not been executed yet.
```

Another great thing about GraphQL Shield is a great built-in caching mechanism. GraphQL AuthZ has no built-in caching at the moment but you could implement it by yourself outside of GraphQL AuthZ or inside of it (pull requests are more than welcome!)

## Conclusion

> Please check out the [GraphQL AuthZ GitHub repo](https://github.com/AstrumU/graphql-authz). You can find a tutorial on how to set it up in different ways and with different technologies. Also, you can find the following working examples:

Reviewer: ```suggestion
GraphQL AuthZ is a new approach to doing GraphQL native authorization. We are happy that we can finally share this library with the community and keen to learn about the ways it might be used within your next project! Don't hesitate to contact us for questions or to evaluate whether this library might be a fit for your architecture.

Please check out the [GraphQL AuthZ repository on GitHub](https://github.com/AstrumU/graphql-authz) for learning more. You can find a tutorial on how to set it up in different ways and with different technologies.

The repository also contains the following, ready to run, examples:
```

The middleware approach of GraphQL shield is missing these possibilities because authorization logic is executed in the context of field resolvers and has access only to the parent object which is the value returned from the parent field resolver. At that stage, the wrapped field resolvers have not been executed yet.

Another feature in GraphQL Shield is the built-in contextual caching mechanism for rules. At the moment, GraphQL AuthZ has no built-in caching, but you can implement it yourself within the GraphQL AuthZ rules, or in the future caching could even become a built-in feature. (pull requests are more than welcome)

## Which approach to choose?

> To sum up use-cases and different authorization approaches, here is the tip:

Reviewer: ```suggestion
Let's wrap up all the mentioned libraries and the use-cases in which they shine:
```

- If you have very few places that should be covered by authorization and you don't plan to increase such places, you can just add authorization logic right inside resolvers to not overcomplicate things.
- If all of your authorization logic doesn't depend on the data and shouldn't perform complex calculations, you can use [operation-field-permissions](https://www.envelop.dev/plugins/use-operation-field-permissions) [Envelop](https://www.envelop.dev/) plugin.
- If your authorization logic contains some calculations which are the same for many fields of your schema, you can use [GraphQL Shield](https://github.com/maticzav/graphql-shield) to leverage the built-in contextual caching mechanism.
- If your authorization logic heavily depends on the data or you want to use schema directives to attach auth rules, you can use [GraphQL-AuthZ](https://github.com/AstrumU/graphql-authz) to leverage the 'GraphQL schema as a data source' pattern and post-execution rules.

## Conclusion

> GraphQL AuthZ is a new approach to doing GraphQL native authorization. We are happy that we can finally share this library with the community and keen to learn about the ways it might be used within your next project! Don't hesitate to contact us for questions or to evaluate whether this library might be a fit for your architecture.

Reviewer: ```suggestion
GraphQL AuthZ is a new approach for applying GraphQL native authorization. We are happy that we can finally share this library with the community and keen to learn about the ways it might be used within your next project! Don't hesitate to contact us for questions or to evaluate whether this library might be a fit for your architecture.
```
